### PR TITLE
Create pastoral record [SA-14709] (master)

### PR DESCRIPTION
**Pull request by tallzeebaa**

### CODE REVIEW

**Review by JustinAClarke — CHANGES_REQUESTED** [review]
some updates please

File: openapi/components/requestBodies/pastoral-item.yaml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+              * `store` for files uploaded to Schoolbox's storage
+              * `browse` finds files in Schoolbox's resource browser
+              * `myfile` finds files in a Schoolbox user's My Files browser
+              * `office365` finds files in an Office365 store
+            * `$location` is the location of the file within `$locator`:
+              * for `store` this is a file's unique hash
+              * for `browse` and `myfile` this is a file ID
+             * for `office365` this is the file's unique node
+            * `$filename` is the filename that the file will be given once it is uploaded to wherever it is being uploaded
+            See the [Upload a file to Schoolbox](#operation/storageAsyncUpload) endpoint for details on how to upload a file, and receive a file pointer from that upload.
```
Comment (JustinAClarke) [inline openapi/components/requestBodies/pastoral-item.yaml:78]: there is already a schema for attaching files, should we be using just referencing that?
`openapi/components/schemas/filePointer.yaml` ??
It looks like it is being used in newsItem.yaml
```
items:
    $ref: ../schemas/filePointer.yaml
```

File: openapi/components/requestBodies/pastoral-item.yaml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        tags:
+          nullable: true
+          description: |
+            Comma separated string of tag names.
+            If Allow Staff to Create Pastoral Care Tags config is ON, attempts to create a new tag, if not alraeady exists.
+          type: string
+        points:
+          nullable: true
+          description: Points of this pastoral record
+          type: integer
```
Comment (JustinAClarke) [inline openapi/components/requestBodies/pastoral-item.yaml:94]: we should limit this to match the front end
`Please enter a value less than or equal to 1000000.`

File: openapi/components/requestBodies/pastoral-item.yaml (hunk 1 above)
Comment (tallzeebaa) [reply openapi/components/requestBodies/pastoral-item.yaml:78]: I tried it first and couldn't make it displayed in the array object.... Does it work for you?

File: openapi/components/requestBodies/pastoral-item.yaml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        role:
+          type: array
+          description: |
+            An array of pairs of a role and its access level to this pastoral record in the format of:
+            `$roleId => $accessLevel`
+            * `$roleId` is either of the followings:
```
Comment (Brahmah) [inline openapi/components/requestBodies/pastoral-item.yaml:100]: > is either of the followings

Typo

File: openapi/components/requestBodies/pastoral-item.yaml (hunk 1 above)
Comment (JustinAClarke) [reply openapi/components/requestBodies/pastoral-item.yaml:78]: same, just tried, it looks like in news it is also specifying the description, and using `$ref` ... weird
lets just leave it here as it. other than updating the link `Upload a file to Schoolbox`

File: openapi/components/requestBodies/pastoral-item.yaml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+              * `store` for files uploaded to Schoolbox's storage
+              * `browse` finds files in Schoolbox's resource browser
+              * `myfile` finds files in a Schoolbox user's My Files browser
+              * `office365` finds files in an Office365 store
+            * `$location` is the location of the file within `$locator`:
+              * for `store` this is a file's unique hash
+              * for `browse` and `myfile` this is a file ID
+             * for `office365` this is the file's unique node
+            * `$filename` is the filename that the file will be given once it is uploaded to wherever it is being uploaded
+            See the [Upload a file to Schoolbox](#operation/storageAsyncUpload) endpoint for details on how to upload a file, and receive a file pointer from that upload.
```
[the remark this answers is not in the document]
Comment (tallzeebaa) [reply openapi/components/requestBodies/pastoral-item.yaml:59]: So.. links in newsItem.yaml is wrong as well. I'll fix them all together.

File: openapi/components/requestBodies/pastoral-item.yaml (hunk 1 above)
Comment (tallzeebaa) [reply openapi/components/requestBodies/pastoral-item.yaml:78]: Actually what 
```
          items:
            $ref: ../schemas/filePointer.yaml
```
in openapi/components/requestBodies/newsItem.yaml is for?

Comment (JustinAClarke) [reply openapi/components/requestBodies/pastoral-item.yaml:78]: it looks like it doesn't really do anything....
image

https://github.com/alaress/schoolbox-api-docs/pull/14/commits/4270fa6782949c60104c7a1d6951f0010399472f

File: openapi/components/schemas/pastoral-item.yaml [hunk 5]
```diff
@@ -0,0 +1,243 @@
+title: a Pastoral Record
+description: A pastoral record
+readOnly: true
+type: array
+properties:
```
Comment (JustinAClarke) [inline openapi/components/schemas/pastoral-item.yaml:5]:
```suggestion
type: array
items:
    properties:
```
it is not displaying the array correctly

File: openapi/paths/pastoral@record@insert.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+post:
+  operationId: pastoral.postRecordInsert
+  tags: [pastoral]
+  summary: Create a new pastoral record
+  description: |
+    Creates a new pastoral record: the record's author will be the currently
+    authenticated user.
```
Comment (JustinAClarke) [inline openapi/paths/pastoral@record@insert.yaml:7]: Can we also please specify the permissions of this endpoint?

File: openapi/components/schemas/pastoral-item.yaml [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        type: string
+      sentiment:
+        type: string
+      sequence:
+        description: The order in which this pastoral type will display in a pastoral type list
+        type: integer
+  subType:
+    description: Sub-type of the pastoral record
+    type: object
+    properties:
```
Comment (JustinAClarke) [inline openapi/components/schemas/pastoral-item.yaml:51]: subtype is nullable, we should specify it here

File: openapi/components/schemas/pastoral-item.yaml [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        type: string
+      sentiment:
+        type: string
+      sequence:
+        description: The order in which this pastoral sub-type will display in a pastoral sub-type list
+        type: integer
+  severity:
+    description: Severity of the pastoral record
+    type: object
+    properties:
```
Comment (JustinAClarke) [inline openapi/components/schemas/pastoral-item.yaml:72]: Severity can also be null

File: openapi/components/schemas/pastoral-item.yaml [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  bodyPrivate:
+    description: |
+      Confidential content of this pastoral record in raw HTML
+    type: string
+  attachmentsPublic:
+    description: |
+      Files attached to General content
+    type: array
+    items:
+      $ref: ./file.yaml
```
Comment (JustinAClarke) [inline openapi/components/schemas/pastoral-item.yaml:111]: in this files object we are also missing a type field:
```
"attachmentsPublic": [
      {
        "hash": "b943b9c127c2c3e3656dd208d2583ec47b7a7c9b",
        "mime": "image\/gif",
        "type": "image",
        "size": 9158,
        "_links": {
          "fetch": "\/storage\/fetch.php?hash=b943b9c127c2c3e3656dd208d2583ec47b7a7c9b",
          "icon": "https:\/\/vagrant.dev.schoolbox.cloud\/static\/22.1.3-0\/images\/mime\/application-image-gif.svg",
          "image": "\/storage\/image.php?hash=b943b9c127c2c3e3656dd208d2583ec47b7a7c9b"
        }
      }
    ]
```

Comment (JustinAClarke) [reply openapi/components/schemas/pastoral-item.yaml:111]: nope type exists, I was cached...

File: openapi/components/schemas/pastoral-item.yaml [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    description: Type of the pastoral record
+    type: object
+    properties:
+      id:
+        description: The ID of the pastoral type associated to this pastoral record
+        type: integer
+      name:
+        description: The name of the pastoral type
+        type: string
+      colour:
```
Comment (JustinAClarke) [inline openapi/components/schemas/pastoral-item.yaml:37]: it doesn't look like we have colours for pastoral types or sub types... is this future planning?

File: openapi/components/requestBodies/pastoral-item.yaml [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            * `$location` is the location of the file within `$locator`:
+              * for `store` this is a file's unique hash
+              * for `browse` and `myfile` this is a file ID
+             * for `office365` this is the file's unique node
+            * `$filename` is the filename that the file will be given once it is uploaded to wherever it is being uploaded
+
+            See the [Upload a file](#post-/storage/asyncUpload.php) endpoint for details on how to upload a file, and receive a file pointer from that upload.
+        actions[]:
+          nullable: true
+          description: An array of IDs of pastoral action
```
Comment (JustinAClarke) [inline openapi/components/requestBodies/pastoral-item.yaml:83]: only available if the config is enabled

File: openapi/components/schemas/pastoral-item.yaml (hunk 10 above)
Comment (tallzeebaa) [reply openapi/components/schemas/pastoral-item.yaml:37]: we do...

Comment (JustinAClarke) [reply openapi/components/schemas/pastoral-item.yaml:37]: I thought it was just for Severity ??

Comment (tallzeebaa) [reply openapi/components/schemas/pastoral-item.yaml:37]: Colour of icons.

**Comment by armatronic** [comment]
@JustinAClarke was there anything further you wanted to add here before this gets merged?

**Comment by JustinAClarke** [comment]
i recall it is all good, and we can merge

**Pull request merged**